File: dydx_v4_client/indexer/rest/modules/markets.py
from typing import Optional

from dydx_v4_client.indexer.rest.constants import TimePeriod
from dydx_v4_client.indexer.rest.shared.rest import RestClient


class MarketsClient(RestClient):

    async def get_perpetual_markets(self, market: Optional[str] = None) -> dict:
        """
        Retrieves perpetual markets.

        Args:
            market (Optional[str]): The specific market ticker to retrieve. If not provided, all markets are returned.

        Returns:
            dict: The perpetual markets data.
        """
        uri = "/v4/perpetualMarkets"
        return await self.get(uri, params={"ticker": market})

    async def get_perpetual_market_orderbook(self, market: str) -> dict:
        """
        Retrieves the orderbook for a specific perpetual market.

        Args:
            market (str): The market ticker.

        Returns:
            dict: The orderbook data.
        """
        uri = f"/v4/orderbooks/perpetualMarket/{market}"
        return await self.get(uri)

    async def get_perpetual_market_trades(
        self,
        market: str,
        starting_before_or_at_height: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> dict:
        """
        Retrieves trades for a specific perpetual market.

        Args:
            market (str): The market ticker.
            starting_before_or_at_height (Optional[int]): The block height to start retrieving trades from.
            limit (Optional[int]): The maximum number of trades to retrieve.

        Returns:
            dict: The trades data.
        """
        uri = f"/v4/trades/perpetualMarket/{market}"
        return await self.get(
            uri,
            params={
                "createdBeforeOrAtHeight": starting_before_or_at_height,
                "limit": limit,
            },
        )

    async def get_perpetual_market_candles(
        self,
        market: str,
        resolution: str,
        from_iso: Optional[str] = None,
        to_iso: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> dict:
        """
        Retrieves candle data for a specific perpetual market.

        Args:
            market (str): The market ticker.
            resolution (str): The candle resolution (e.g., "1DAY", "1HOUR", "1MIN").
            from_iso (Optional[str]): The start timestamp in ISO format.
            to_iso (Optional[str]): The end timestamp in ISO format.
            limit (Optional[int]): The maximum number of candles to retrieve.

        Returns:
            dict: The candle data.
        """
        uri = f"/v4/candles/perpetualMarkets/{market}"
        return await self.get(
            uri,
            params={
                "resolution": resolution,
                "fromISO": from_iso,
                "toISO": to_iso,
                "limit": limit,
            },
        )

    async def get_perpetual_market_historical_funding(
        self,
        market: str,
        effective_before_or_at: Optional[str] = None,
        effective_before_or_at_height: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> dict:
        """
        Retrieves historical funding rates for a specific perpetual market.

        Args:
            market (str): The market ticker.
            effective_before_or_at (Optional[str]): The timestamp to retrieve funding rates effective before or at.
            effective_before_or_at_height (Optional[int]): The block height to retrieve funding rates effective before or at.
            limit (Optional[int]): The maximum number of funding rates to retrieve.

        Returns:
            dict: The historical funding rates data.
        """
        uri = f"/v4/historicalFunding/{market}"
        return await self.get(
            uri,
            params={
                "effectiveBeforeOrAt": effective_before_or_at,
                "effectiveBeforeOrAtHeight": effective_before_or_at_height,
                "limit": limit,
            },
        )

    async def get_perpetual_market_sparklines(
        self, period: str = TimePeriod.ONE_DAY
    ) -> dict:
        """
        Retrieves sparkline data for perpetual markets.

        Args:
            period (str): The time period for the sparkline data (e.g., "ONE_DAY", "SEVEN_DAYS").

        Returns:
            dict: The sparkline data.
        """
        uri = "/v4/sparklines"
        return await self.get(uri, params={"timePeriod": period})

File: v4-client-rs/client/src/indexer/rest/client/markets.rs
use super::*;
use anyhow::{anyhow as err, Error};
use std::collections::HashMap;

/// Markets dispatcher.
///
/// Check [the example](https://github.com/dydxprotocol/v4-clients/blob/main/v4-client-rs/client/examples/market_endpoint.rs).
pub struct Markets<'a> {
    rest: &'a RestClient,
}

impl<'a> Markets<'a> {
    /// Create a new markets dispatcher.
    pub(crate) fn new(rest: &'a RestClient) -> Self {
        Self { rest }
    }

    /// Query for perpetual markets data.
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-perpetual-markets).
    #[deprecated(since = "0.3.0", note = "Use `get_perpetual_markets` instead")]
    pub async fn list_perpetual_markets(
        &self,
        opts: Option<ListPerpetualMarketsOpts>,
    ) -> Result<HashMap<Ticker, PerpetualMarket>, Error> {
        self.get_perpetual_markets(opts).await
    }

    /// Query for perpetual markets data.
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-perpetual-markets).
    pub async fn get_perpetual_markets(
        &self,
        opts: Option<ListPerpetualMarketsOpts>,
    ) -> Result<HashMap<Ticker, PerpetualMarket>, Error> {
        let rest = &self.rest;
        const URI: &str = "/v4/perpetualMarkets";
        let url = format!("{}{URI}", rest.config.endpoint);
        let options = opts.unwrap_or_default();
        let markets = rest
            .client
            .get(url)
            .query(&options)
            .send()
            .await?
            .error_for_status()?
            .json::<PerpetualMarketResponse>()
            .await?
            .markets;
        Ok(markets)
    }

    /// Query for the perpetual market.
    pub async fn get_perpetual_market(&self, ticker: &Ticker) -> Result<PerpetualMarket, Error> {
        let mut markets = self
            .get_perpetual_markets(Some(ListPerpetualMarketsOpts {
                limit: Some(1),
                ticker: Some(ticker.clone()),
            }))
            .await?;
        markets
            .remove(ticker)
            .ok_or_else(|| err!("Market ticker not found in list Markets response"))
    }

    /// Query for bids-asks for the perpetual market.
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-perpetual-market-orderbook).
    pub async fn get_perpetual_market_orderbook(
        &self,
        ticker: &Ticker,
    ) -> Result<OrderBookResponseObject, Error> {
        let rest = &self.rest;
        const URI: &str = "/v4/orderbooks/perpetualMarket";
        let url = format!("{}{URI}/{ticker}", rest.config.endpoint);
        let orderbook = rest
            .client
            .get(url)
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;
        Ok(orderbook)
    }

    /// Query for trades.
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-trades).
    #[deprecated(since = "0.3.0", note = "Use `get_perpetual_market_trades` instead")]
    pub async fn get_trades(
        &self,
        ticker: &Ticker,
        opts: Option<GetTradesOpts>,
    ) -> Result<Vec<TradeResponseObject>, Error> {
        self.get_perpetual_market_trades(ticker, opts).await
    }

    /// Query for trades.
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-trades).
    pub async fn get_perpetual_market_trades(
        &self,
        ticker: &Ticker,
        opts: Option<GetTradesOpts>,
    ) -> Result<Vec<TradeResponseObject>, Error> {
        let rest = &self.rest;
        const URI: &str = "/v4/trades/perpetualMarket";
        let url = format!("{}{URI}/{ticker}", rest.config.endpoint);
        let options = opts.unwrap_or_default();
        let trades = rest
            .client
            .get(url)
            .query(&options)
            .send()
            .await?
            .error_for_status()?
            .json::<TradeResponse>()
            .await?
            .trades;
        Ok(trades)
    }

    /// Query for [candles](https://dydx.exchange/crypto-learning/candlestick-patterns).
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-candles).
    #[deprecated(since = "0.3.0", note = "Use `get_perpetual_market_candles` instead")]
    pub async fn get_candles(
        &self,
        ticker: &Ticker,
        res: CandleResolution,
        opts: Option<GetCandlesOpts>,
    ) -> Result<Vec<CandleResponseObject>, Error> {
        self.get_perpetual_market_candles(ticker, res, opts).await
    }

    /// Query for [candles](https://dydx.exchange/crypto-learning/candlestick-patterns).
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-candles).
    pub async fn get_perpetual_market_candles(
        &self,
        ticker: &Ticker,
        res: CandleResolution,
        opts: Option<GetCandlesOpts>,
    ) -> Result<Vec<CandleResponseObject>, Error> {
        let rest = &self.rest;
        const URI: &str = "/v4/candles/perpetualMarkets";
        let url = format!("{}{URI}/{ticker}", rest.config.endpoint);
        let options = opts.unwrap_or_default();
        let candles = rest
            .client
            .get(url)
            .query(&[("resolution", &res)])
            .query(&options)
            .send()
            .await?
            .error_for_status()?
            .json::<CandleResponse>()
            .await?
            .candles;
        Ok(candles)
    }

    /// Query for funding till time/block specified.
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-historical-funding).
    #[deprecated(
        since = "0.3.0",
        note = "Use `get_perpetual_market_historical_funding` instead"
    )]
    pub async fn get_historical_funding(
        &self,
        ticker: &Ticker,
        opts: Option<GetHistoricalFundingOpts>,
    ) -> Result<Vec<HistoricalFundingResponseObject>, Error> {
        self.get_perpetual_market_historical_funding(ticker, opts)
            .await
    }

    /// Query for funding till time/block specified.
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-historical-funding).
    pub async fn get_perpetual_market_historical_funding(
        &self,
        ticker: &Ticker,
        opts: Option<GetHistoricalFundingOpts>,
    ) -> Result<Vec<HistoricalFundingResponseObject>, Error> {
        let rest = &self.rest;
        const URI: &str = "/v4/historicalFunding";
        let url = format!("{}{URI}/{ticker}", rest.config.endpoint);
        let options = opts.unwrap_or_default();
        let funding = rest
            .client
            .get(url)
            .query(&options)
            .send()
            .await?
            .error_for_status()?
            .json::<HistoricalFundingResponse>()
            .await?
            .historical_funding;
        Ok(funding)
    }

    /// Query for [sparklines](https://en.wikipedia.org/wiki/Sparkline).
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-sparklines).
    #[deprecated(
        since = "0.3.0",
        note = "Use `get_perpetual_market_sparklines` instead"
    )]
    pub async fn get_sparklines(
        &self,
        period: SparklineTimePeriod,
    ) -> Result<SparklineResponseObject, Error> {
        self.get_perpetual_market_sparklines(period).await
    }

    /// Query for [sparklines](https://en.wikipedia.org/wiki/Sparkline).
    ///
    /// [Reference](https://docs.dydx.xyz/indexer-client/http#get-sparklines).
    pub async fn get_perpetual_market_sparklines(
        &self,
        period: SparklineTimePeriod,
    ) -> Result<SparklineResponseObject, Error> {
        let rest = &self.rest;
        const URI: &str = "/v4/sparklines";
        let url = format!("{}{URI}", rest.config.endpoint);
        let sparklines = rest
            .client
            .get(url)
            .query(&[("timePeriod", &period)])
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;
        Ok(sparklines)
    }
}
